File: src/main/java/com/brett/lec08/Lec08Main.java
package com.brett.lec08;

public class Lec08Main {
    public static void main(String[] args) {
//        System.out.println("hello lec08Main java");
        /**
         * 배열에 넣거나
         */
        String[] array = new String[]{"A","B","C"};
        printAll(array);

        /**
         * , 를 통해 직접 넣음
         */
        printAll("D","E","F");
    }

    /**
     * 여러개의 String 문자열을 들어올 때 처리된다.
     * String...
     * @param strings
     */
    public static void printAll(String... strings){
        for(String str : strings){
            System.out.println(str);
        }
    }

    public static void repeat(String str, int num, boolean useNewLine){
        for(int i=1;i<=num;i++){
            if(useNewLine){
                System.out.println(str);
            }else{
                System.out.print(str);
            }
        }
    }

    private static void repeat(String str, int num){
        repeat(str, num, true);
    }

    private static void repeat(String str){
        repeat(str, 3, true);
    }

    private static int max(int a, int b){
        if(a>b){
            return a;
        }else{
            return b;
        }
    }
}

File: src/main/kotlin/com/brett/lec08/Lec08Main.kt
package com.brett.lec08

fun main(){
//    println("hello lec08Main kotlin")
//    println(max_V2(1,2))
//    repeat("hello world", 3, true)
//    repeat("hello world V2")

    // default parameter
    /**
     * 직접 3개의 파라미터를 넣어 선언해주는 경우
     */
//    repeat("hello world V3", 3, false)
    /**
     * 매개변수 이름을 통해 직접 지정
     * 지정되지 않은 매개변수는 기본값 사용
     */
//    repeat("hello world V2", useNewLine = false)

    // named Argument
    printNameAndGender(name = "방정환", gender = "MALE")

//    Lec08Main.repeat(str="name",) // -> Named arguments are not allowed for non-Kotlin functions

    printAll("A","B","C")
    /**
     * 배열을 넣을 때에는 *(spread 연산자)를 사용해야함
     */
    val array = arrayOf("D","E","F")
    printAll(*array)

}

// 같은 타입의 여러 파라미터 받기 (가변인자)
fun printAll(vararg strings : String){
    for(str in strings){
        println(str)
    }
}

// named Argument
fun printNameAndGender(name : String, gender : String){
    println(name)
    println(gender)
}

// default parameter
/**
 * 밖에서 파라미터를 넣어주지 않는 경우 기본값을 사용
 */
fun repeat(
    str: String?,
    num: Int = 3,
    useNewLine: Boolean = true
) {
    for (i in 1..num) {
        if (useNewLine) {
            println(str)
        } else {
            print(str)
        }
    }
}

fun max_V1(a : Int, b : Int) : Int{
    if (a > b) {
        return a
    } else {
        return b
    }
}

/**
 * if-else expression
 */
fun max_V2(a : Int, b : Int) : Int{
    return if (a > b) {
        a
    } else {
        b
    }
}

/**
 * 함수의 결과가 하나면 block({}) 대신 = 사용 가능
 */
fun max_V3(a: Int, b: Int) : Int =
    if (a > b) {
        a
    } else {
        b
    }

/**
 * 한줄로 변경 가능
 * = 을 썼기 때문에 반환 타입 생략 가능
 * block({}) 사용하는 경우에는 반환타입이 Unit이 아니라면 명시적으로 작성 필요
 */
fun max_V4(a: Int, b: Int) = if (a > b) a else b

